xrootd Auth failed in singularity on Dirac job

Accessing root grid file through xrootd failed on a dirac job when using singularity virtualisation. Now singularity is not expected to be found in `$PATH` on any dirac site had to use ganga pr #1670 to use singularity binary from cvmfs. However, i get an authentication failure when i try to access the root file with xrootd. Here's the simple shell script i'm running:
```shell
source /cvmfs/sft.cern.ch/lcg/views/LCG_95apython3/x86_64-centos7-gcc8-opt/setup.sh
python -c "import ROOT; ROOT.TFile.Open('root://gfe02.grid.hep.ph.ic.ac.uk/pnfs/hep.ph.ic.ac.uk/data/lz/lz/data/MDC3/calibration/LZAP-4.3.1/20180201/lz_20180201234_lzap.root').Get('Events')"
```
This is the ganga python job:
```python
    j = ganga.Job()
    j.application=Executable(exe=File('worker_node_script.sh'), args=[])
    j.virtualization = Singularity("docker://luxzeplin/offline_hosted:centos7_2")
    j.virtualization.binary='/cvmfs/oasis.opensciencegrid.org/mis/singularity/current/bin/singularity'
    j.virtualization.mounts = {'/cvmfs':'/cvmfs', '/srv':'/srv'}
    j.backend = Dirac()
    j.submit()
```
This is the error message:
```python
Error in <TNetXNGFile::Open>: [FATAL] Auth failed
Traceback (most recent call last):
  File "<string>", line 1, in <module>
ReferenceError: attempt to access a null-pointer
```
Here are the x509 envs for the dirac singularity:
```shell
printenv | grep SINGULARITY
SINGULARITY_APPNAME=
SINGULARITY_NAME=singularity_sandbox
SINGULARITY_CONTAINER=/tmp/25355128/singularity_sandbox
printenv | grep X509
X509_CERT_DIR=/etc/grid-security/certificates
X509_VOMS_DIR=/tmp/etc/grid-security/vomsdir
X509_USER_PROXY=/tmp/proxy
```
Within the container you can access `X509_VOMS_DIR`, `X509_USER_PROXY` but not `X509_CERT_DIR`. 

When re-running and setting `export XRD_LOGLEVEL="Debug"` here is the log:
```shell
[2020-05-16 16:53:22.355183 +0000][Warning][Utility           ] Unable to process global config file: [ERROR] OS Error: No such file or directory
[2020-05-16 16:53:22.355298 +0000][Debug  ][Utility           ] Unable to process user config file: [ERROR] OS Error: No such file or directory
[2020-05-16 16:53:22.358257 +0000][Debug  ][PlugInMgr         ] Initializing plug-in manager...
[2020-05-16 16:53:22.358284 +0000][Debug  ][PlugInMgr         ] No default plug-in, loading plug-in configs...
[2020-05-16 16:53:22.358293 +0000][Debug  ][PlugInMgr         ] Processing plug-in definitions in /etc/xrootd/client.plugins.d...
[2020-05-16 16:53:22.358311 +0000][Debug  ][PlugInMgr         ] Unable to process directory /etc/xrootd/client.plugins.d: [ERROR] OS Error: No such file or directory
[2020-05-16 16:53:22.358353 +0000][Debug  ][PlugInMgr         ] Processing plug-in definitions in /mnt/dirac/.xrootd/client.plugins.d...
[2020-05-16 16:53:22.358365 +0000][Debug  ][PlugInMgr         ] Unable to process directory /mnt/dirac/.xrootd/client.plugins.d: [ERROR] OS Error: No such file or directory
[2020-05-16 16:53:22.492607 +0000][Debug  ][Poller            ] Available pollers: built-in
[2020-05-16 16:53:22.492897 +0000][Debug  ][Poller            ] Attempting to create a poller according to preference: built-in
[2020-05-16 16:53:22.492905 +0000][Debug  ][Poller            ] Creating poller: built-in
[2020-05-16 16:53:22.492923 +0000][Debug  ][Poller            ] Creating and starting the built-in poller...
[2020-05-16 16:53:22.493389 +0000][Debug  ][Poller            ] Using 1 poller threads
[2020-05-16 16:53:22.493407 +0000][Debug  ][TaskMgr           ] Starting the task manager...
[2020-05-16 16:53:22.493441 +0000][Debug  ][TaskMgr           ] Task manager started
[2020-05-16 16:53:22.493449 +0000][Debug  ][JobMgr            ] Starting the job manager...
[2020-05-16 16:53:22.493526 +0000][Debug  ][JobMgr            ] Job manager started, 3 workers
[2020-05-16 16:53:22.493540 +0000][Debug  ][TaskMgr           ] Registering task: "FileTimer task" to be run at: [2020-05-16 16:53:22 +0000]
[2020-05-16 16:53:22.493617 +0000][Debug  ][Utility           ] Env: overriding entry: MultiProtocol=0 with 1
[2020-05-16 16:53:22.493735 +0000][Debug  ][File              ] [0x3cd9980@root://gfe02.grid.hep.ph.ic.ac.uk:1094/pnfs/hep.ph.ic.ac.uk/data/lz/lz/data/MDC3/calibration/LZAP-4.3.1/20180201/lz_20180201234_lzap.root] Sending an open command
[2020-05-16 16:53:22.493822 +0000][Debug  ][PostMaster        ] Creating new channel to: gfe02.grid.hep.ph.ic.ac.uk:1094 1 stream(s)
[2020-05-16 16:53:22.493849 +0000][Debug  ][PostMaster        ] [gfe02.grid.hep.ph.ic.ac.uk:1094 #0] Stream parameters: Network Stack: IPAuto, Connection Window: 120, ConnectionRetry: 5, Stream Error Widnow: 1800
[2020-05-16 16:53:22.498436 +0000][Debug  ][TaskMgr           ] Registering task: "TickGeneratorTask for: gfe02.grid.hep.ph.ic.ac.uk:1094" to be run at: [2020-05-16 16:53:37 +0000]
[2020-05-16 16:53:22.500402 +0000][Debug  ][PostMaster        ] [gfe02.grid.hep.ph.ic.ac.uk:1094] Found 1 address(es): [::ffff:146.179.232.84]:1094
[2020-05-16 16:53:22.500457 +0000][Debug  ][AsyncSock         ] [gfe02.grid.hep.ph.ic.ac.uk:1094 #0.0] Attempting connection to [::ffff:146.179.232.84]:1094
[2020-05-16 16:53:22.500738 +0000][Debug  ][Poller            ] Adding socket 0x4a0aad0 to the poller
[2020-05-16 16:53:22.503441 +0000][Debug  ][AsyncSock         ] [gfe02.grid.hep.ph.ic.ac.uk:1094 #0.0] Async connection call returned
[2020-05-16 16:53:22.503494 +0000][Debug  ][XRootDTransport   ] [gfe02.grid.hep.ph.ic.ac.uk:1094 #0.0] Sending out the initial hand shake + kXR_protocol
[2020-05-16 16:53:22.503987 +0000][Debug  ][XRootDTransport   ] [gfe02.grid.hep.ph.ic.ac.uk:1094 #0.0] Got the server hand shake response (type: manager [], protocol version 400)
[2020-05-16 16:53:22.504594 +0000][Debug  ][XRootDTransport   ] [gfe02.grid.hep.ph.ic.ac.uk:1094 #0.0] kXR_protocol successful (type: server [], protocol version 400)
[2020-05-16 16:53:22.508272 +0000][Debug  ][XRootDTransport   ] [gfe02.grid.hep.ph.ic.ac.uk:1094 #0.0] Sending out kXR_login request, username: dirac, cgi: ?xrd.cc=us&xrd.tz=0&xrd.appname=python3.6&xrd.info=&xrd.hostname=dirac-26a1e6ef.novalocal&xrd.rn=v4.8.4, dual-stack: true, private IPv4: true, private IPv6: true
[2020-05-16 16:53:22.534402 +0000][Debug  ][XRootDTransport   ] [gfe02.grid.hep.ph.ic.ac.uk:1094 #0.0] Logged in, session: 7e78f76467065d93ac52088c47155129
[2020-05-16 16:53:22.534417 +0000][Debug  ][XRootDTransport   ] [gfe02.grid.hep.ph.ic.ac.uk:1094 #0.0] Authentication is required: &P=gsi,v:10400,c:ssl,ca:ffc3d59b
[2020-05-16 16:53:22.534424 +0000][Debug  ][XRootDTransport   ] [gfe02.grid.hep.ph.ic.ac.uk:1094 #0.0] Sending authentication data
200516 16:53:22 447 secgsi_Init: Secgsi: ErrError: CA directory non existing:: /etc/grid-security/certificates
200516 16:53:22 447 secgsi_Init: Secgsi: ErrError: CRL directory non existing:: /etc/grid-security/certificates
[2020-05-16 16:53:22.636886 +0000][Debug  ][XRootDTransport   ] [gfe02.grid.hep.ph.ic.ac.uk:1094 #0.0] Trying to authenticate using gsi
[2020-05-16 16:53:22.652224 +0000][Debug  ][XRootDTransport   ] [gfe02.grid.hep.ph.ic.ac.uk:1094 #0.0] Cannot get credentials for protocol gsi: Secgsi: ErrParseBuffer: unknown CA: cannot verify server certificate: kXGS_init
[2020-05-16 16:53:22.652246 +0000][Error  ][XRootDTransport   ] [gfe02.grid.hep.ph.ic.ac.uk:1094 #0.0] No protocols left to try
[2020-05-16 16:53:22.652260 +0000][Error  ][AsyncSock         ] [gfe02.grid.hep.ph.ic.ac.uk:1094 #0.0] Socket error while handshaking: [FATAL] Auth failed
[2020-05-16 16:53:22.652269 +0000][Debug  ][AsyncSock         ] [gfe02.grid.hep.ph.ic.ac.uk:1094 #0.0] Closing the socket
[2020-05-16 16:53:22.652281 +0000][Debug  ][Poller            ] <[::ffff:172.16.1.164]:54048><--><[::ffff:146.179.232.84]:1094> Removing socket from the poller
[2020-05-16 16:53:22.652325 +0000][Error  ][PostMaster        ] [gfe02.grid.hep.ph.ic.ac.uk:1094 #0] elapsed = 0, pConnectionWindow = 120 seconds.
[2020-05-16 16:53:22.652333 +0000][Error  ][PostMaster        ] [gfe02.grid.hep.ph.ic.ac.uk:1094 #0] Unable to recover: [FATAL] Auth failed.
[2020-05-16 16:53:22.652346 +0000][Error  ][XRootD            ] [gfe02.grid.hep.ph.ic.ac.uk:1094] Impossible to send message kXR_open (file: pnfs/hep.ph.ic.ac.uk/data/lz/lz/data/MDC3/calibration/LZAP-4.3.1/20180201/lz_20180201234_lzap.root, mode: 00, flags: kXR_open_read kXR_async kXR_retstat ). Trying to recover.
[2020-05-16 16:53:22.652356 +0000][Debug  ][XRootD            ] [gfe02.grid.hep.ph.ic.ac.uk:1094] Handling error while processing kXR_open (file: pnfs/hep.ph.ic.ac.uk/data/lz/lz/data/MDC3/calibration/LZAP-4.3.1/20180201/lz_20180201234_lzap.root, mode: 00, flags: kXR_open_read kXR_async kXR_retstat ): [FATAL] Auth failed.
[2020-05-16 16:53:22.655048 +0000][Debug  ][File              ] [0x3cd9980@root://gfe02.grid.hep.ph.ic.ac.uk:1094/pnfs/hep.ph.ic.ac.uk/data/lz/lz/data/MDC3/calibration/LZAP-4.3.1/20180201/lz_20180201234_lzap.root] Open has returned with status [FATAL] Auth failed
[2020-05-16 16:53:22.655062 +0000][Debug  ][File              ] [0x3cd9980@root://gfe02.grid.hep.ph.ic.ac.uk:1094/pnfs/hep.ph.ic.ac.uk/data/lz/lz/data/MDC3/calibration/LZAP-4.3.1/20180201/lz_20180201234_lzap.root] Error while opening at gfe02.grid.hep.ph.ic.ac.uk:1094: [FATAL] Auth failed
[2020-05-16 16:53:22.655085 +0000][Debug  ][Utility           ] Monitor library name not set. No monitoring
Error in <TNetXNGFile::Open>: [FATAL] Auth failed
Traceback (most recent call last):
  File "<string>", line 1, in <module>
ReferenceError: attempt to access a null-pointer
[2020-05-16 16:53:22.669386 +0000][Debug  ][JobMgr            ] Stopping the job manager...
[2020-05-16 16:53:22.670221 +0000][Debug  ][JobMgr            ] Job manager stopped
[2020-05-16 16:53:22.670242 +0000][Debug  ][TaskMgr           ] Stopping the task manager...
[2020-05-16 16:53:22.670376 +0000][Debug  ][TaskMgr           ] Task manager stopped
[2020-05-16 16:53:22.670387 +0000][Debug  ][Poller            ] Stopping the poller...
[2020-05-16 16:53:22.670943 +0000][Debug  ][TaskMgr           ] Requesting unregistration of: "TickGeneratorTask for: gfe02.grid.hep.ph.ic.ac.uk:1094"
[2020-05-16 16:53:22.670962 +0000][Debug  ][AsyncSock         ] [gfe02.grid.hep.ph.ic.ac.uk:1094 #0.0] Closing the socket
[2020-05-16 16:53:22.670969 +0000][Debug  ][PostMaster        ] [gfe02.grid.hep.ph.ic.ac.uk:1094 #0] Destroying stream
[2020-05-16 16:53:22.670978 +0000][Debug  ][AsyncSock         ] [gfe02.grid.hep.ph.ic.ac.uk:1094 #0.0] Closing the socket
```

Similar issue to #1668

--- a/ganga/GangaCore/Lib/Virtualization/Singularity.py
+++ b/ganga/GangaCore/Lib/Virtualization/Singularity.py
@@ -99,7 +99,10 @@ class Singularity(IVirtualization):
 print("Using singularity")
 
 import stat
-if not ( ('XDG_RUNTIME_DIR' in runenv) and os.path.isdir(runenv['XDG_RUNTIME_DIR']) and (stat.S_IMODE(os.stat(runenv['XDG_RUNTIME_DIR']).st_mode) == 0o700) ):
+if not ( ('XDG_RUNTIME_DIR' in runenv) and
+         os.path.isdir(runenv['XDG_RUNTIME_DIR']) and
+         (stat.S_IMODE(os.stat(runenv['XDG_RUNTIME_DIR']).st_mode) == 0o700) and
+         os.access(runenv['XDG_RUNTIME_DIR'], os.W_OK) ):
     os.mkdir('.xdg', 0o700)
     runenv['XDG_RUNTIME_DIR'] = os.path.join(os.getcwd(), '.xdg')
            

--- a/ganga/GangaDirac/Lib/RTHandlers/ExeDiracRTHandler.py
+++ b/ganga/GangaDirac/Lib/RTHandlers/ExeDiracRTHandler.py
@@ -181,6 +181,12 @@ print("files on WN: " + str(listdir('.')))
 
 ###OUTPUTFILESINJECTEDCODE###
 
+try:
+    subprocess.call('chmod -R u+rwX  * .*', env=runenv, shell=True)
+except Exception as x:
+    print('Problems fixing permissions on output files to allow for clean-up.')
+    print('Err was: ' + str(x))
+
 sys.exit(rc)
 """
     return script_template

--- a/ganga/GangaLHCb/Lib/RTHandlers/ExeDiracRTHandler.py
+++ b/ganga/GangaLHCb/Lib/RTHandlers/ExeDiracRTHandler.py
@@ -192,6 +192,12 @@ print("files on WN: " + str(listdir('.')))
 
 ###OUTPUTFILESINJECTEDCODE###
 
+try:
+    subprocess.call('chmod -R u+rwX  * .*', env=runenv, shell=True)
+except Exception as x:
+    print('Problems fixing permissions on output files to allow for clean-up.')
+    print('Err was: ' + str(x))
+
 sys.exit(rc)
 """
     return script_template

